src/Http/Http.Extensions/src/RequestDelegateBuilder.cs

```diff
@@ -44,7 +46,86 @@ internal static class MapActionExpressionTreeBuilder
+            }
+
+            var targetExpression = action.Target switch
+            {
+                { } => Expression.Convert(TargetArg, action.Target.GetType()),
+                null => null,
+            };
+
+            var untargetedRequestDelegate = BuildRequestDelegate(action.Method, targetExpression);
+
+            return httpContext =>
+            {
+                return untargetedRequestDelegate(action.Target, httpContext);
+            };
+        }
+
+        /// <summary>
+        /// Builds a <see cref="RequestDelegate"/> implementation for <paramref name="methodInfo"/>.
+        /// </summary>
+        /// <param name="methodInfo">A static request handler with any number of custom parameters that often produces a response with its return value.</param>
+        /// <returns>The <see cref="RequestDelegate"/></returns>
+        public static RequestDelegate BuildRequestDelegate(MethodInfo methodInfo)
+        {
+            if (methodInfo is null)
+            {
+                throw new ArgumentNullException(nameof(methodInfo));
+            }
+
+            var untargetedRequestDelegate = BuildRequestDelegate(methodInfo, targetExpression: null);
+
+            return httpContext =>
+            {
+                return untargetedRequestDelegate(null, httpContext);
+            };
+        }
+
+
+        /// <summary>
+        /// Builds a <see cref="RequestDelegate"/> implementation for <paramref name="methodInfo"/>.
+        /// </summary>
+        /// <param name="methodInfo">A request handler with any number of custom parameters that often produces a response with its return value.</param>
+        /// <param name="targetFactory">Creates the <see langword="this"/> for the non-static method. If the </param>
+        /// <returns>The <see cref="RequestDelegate"/></returns>
+        public static RequestDelegate BuildRequestDelegate(MethodInfo methodInfo, Func<HttpContext, object> targetFactory)
+        {
+            if (methodInfo is null)
+            {
+                throw new ArgumentNullException(nameof(methodInfo));
+            }
+
+            if (targetFactory is null)
+            {
+                throw new ArgumentNullException(nameof(targetFactory));
+            }
+
+            if (methodInfo.DeclaringType is null)
+            {
+                throw new ArgumentException($"A {nameof(targetFactory)} was provided, but {nameof(methodInfo)} does not have a Declaring type.");
+            }
+
+            var targetExpression = Expression.Convert(TargetArg, methodInfo.DeclaringType);
+            var untargetedRequestDelegate = BuildRequestDelegate(methodInfo, targetExpression);
+
+            return httpContext =>
+            {
+                return untargetedRequestDelegate(targetFactory(httpContext), httpContext);
+            };
+        }
+
+        private static Func<object?, HttpContext, Task> BuildRequestDelegate(MethodInfo methodInfo, Expression? targetExpression)
```

Comment:
Though this was easier, I'd like to see this not allocate an extra delegate when it doesn't need to.

Reply:
Though, I'd be fine changing that when we do more performance work.

Reply:
I guess a thing to note in these APIs would be that this particular RequestDelegate has very specific behavior around how the input is processed i.e. it has MapAction semantics. 

Would developers want to change this? For instance, would we build this API in a way that you could build MVC on top of it? If not, we really should include much more details about how the constructed RequestDelegate behaves as part of the docs.

Reply:
Agreed this kind of optimization should be saved for later.

Remember, we ultimately have to reference this extra "target" parameter that can change per invocation that isn't normally an argument taken by `RequestDelegate`, so fixing this isn't trivial. Removing the allocation would have to involve something like setting it on an HttpContext field (via an optimized internal Feature perhaps) or a pool of RequestDelegates that have an indirect reference to the ever-changing "target"

In any case, the fix isn't some trivial improvement to the Expression tree building. The previous code also allocated this "extra" delegate so this isn't a regression at all.

Reply:
Oh yeah. And this extra allocation only happens during the actual call to Build(), not every invocation. Not saying build shouldn't be fast. I understand people might want to call it a lot, but in the context of Expression tree building, I think the optimization is a drop in the bucket.

Reply:
We're gonna rewrite this all to use ref emit 🤣.

Reply:
🤣 And even if that was fast there's also the reflection.